File: Problems/hacker_rank/Arrays_and_Sorting/20150105_CountingSort3/Solution.scala
object Solution {
  def countingSort(ar: Array[Int]): Array[Int] = {
    val counts: Array[Int] = Array.fill[Int](100)(0)
    for ( a <- ar ) {
      counts(a) += 1
    }
    for ( i <- 1 until 100 )  {
      if ( 0 < counts(i) )  {
        counts(i) += counts(i - 1)
      } else if ( 0 == counts(i) )  {
        counts(i) = counts(i - 1)
      }
    }
    counts
  }

  def main(args: Array[String]) {
    val inps = io.Source.stdin.getLines.drop(1).toArray
    val ar = inps.map((s: String) => s.split(" ")(0).toInt).toArray
    val res = countingSort(ar)
    println(res.mkString(" "))
  }
}

File: Problems/hacker_rank/Arrays_and_Sorting/20150105_CountingSort3/Solution.java
import java.util.Scanner;

public class Solution {
  public static int[] countingSort(final int[] ar)  {
    int[] counts = new int[100];
    for ( int i = 0; i < ar.length; ++i )  {
      ++counts[ar[i]];
    }
    for ( int i = 1; i < 100; ++i ) {
      if ( 0 < counts[i] )  {
        counts[i] += counts[i - 1];
      } else if ( 0 == counts[i] )  {
        counts[i] = counts[i - 1];
      }
    }
    return counts;
  }

  public static void main(final String[] args)  {
    Scanner sc = new Scanner(System.in);
    final int n = Integer.parseInt(sc.nextLine());
    int[] ar = new int[n];
    for ( int i = 0; i < n; ++i ) {
      String[] inps = sc.nextLine().split(" ");
      ar[i] = Integer.parseInt(inps[0]);
    }
    int[] res = countingSort(ar);
    System.out.print(res[0]);
    for ( int i = 1; i < res.length; ++i )  {
      System.out.print(" " + res[i]);
    }
    System.out.println();
  }
}
